parser.js

```diff
@@ -0,0 +1,113 @@
+import { RootNode, Node, TextNode } from './class.js';
+
+const removeLineBreaks = (str) => str.replace(/(\r\n|\n|\r)/gm, '');
+
+/* @params attributesToBeParsed : ['class="container"'] */
+/* @return attributes : [['class', 'container']] */
+const parseAttributes = (attributesToBeParsed) => {
+  return attributesToBeParsed.map((v) => {
+    let [key, value] = v.split('=');
+    if (!value) {
+      value = true;
+    } else {
+      value = value.replace(/"/g, '');
+    }
+    return [key, value];
+  });
+};
+
+const isOpeningTag = (t, index) => t[index] === '<' && t[index + 1] !== '/';
+
+const getIndexOfClosingBracket = (t, indexOpeningTagStart) => {
+  let indexClosingBracket = indexOpeningTagStart;
+
+  while (t[indexClosingBracket] !== '>') {
+    indexClosingBracket++;
+  }
+  return indexClosingBracket;
+};
+
+const getIndexOfTextEnd = (t, indexOfTextStart) => {
+  let indexTextEnd = indexOfTextStart;
+
+  while (t[indexTextEnd] !== '<' && indexTextEnd < t.length) {
+    indexTextEnd++;
+  }
+  return indexTextEnd - 1;
+};
+
+const getIndexOfClosingTag = (t, tagName, indexOpeningTagStart) => {
+  const openingTag = `<${tagName}`;
+  const closingTag = `</${tagName}>`;
+  const openingIndexes = [...t.matchAll(new RegExp(openingTag, 'gi'))].map((v) => v.index);
+  const closingIndexes = [...t.matchAll(new RegExp(closingTag, 'gi'))].map((v) => v.index);
+  let op = 0;
+  let cp = 0;
+  let accCount = 0;
+
+  for (let i = 0; i < t.length; i++) {
+    if (i === openingIndexes[op]) {
+      accCount += 1;
+      op++;
+    } else if (i === closingIndexes[cp]) {
+      accCount -= 1;
+
+      if (i > indexOpeningTagStart && accCount === 0) {
+        return closingIndexes[cp];
+      }
+      cp++;
+    }
+  }
+  return closingIndexes.slice(-1)[0];
+};
+
+const getNodeInfo = (t, indexStart) => {
+  const indexOpeningTagEnd = getIndexOfClosingBracket(t, indexStart);
+  const chunk = t.slice(indexStart + 1, indexOpeningTagEnd);
+  const [tagName, ...attributesToBeParsed] = chunk.split(' ');
+  const attributes = parseAttributes(attributesToBeParsed);
+  const indexClosingTagStart = getIndexOfClosingTag(t, tagName, indexStart);
+  const indexClosingTagEnd = indexClosingTagStart + `/${tagName}>`.length;
+  const childrenTemplate = t.slice(indexOpeningTagEnd + 1, indexClosingTagStart);
+
+  return { tagName, attributes, indexClosingTagEnd, childrenTemplate };
+};
+
+const getTextNodeInfo = (t, indexStart) => {
+  const indexTextEnd = getIndexOfTextEnd(t, indexStart);
+  const value = t.slice(indexStart, indexTextEnd + 1);
+
+  return { value: value.trim(), indexTextEnd };
+};
+
+/* @params template : string */
+/* @params parent : RootNode | TextNode */
+/* @return parent : RootNode | TextNode (children updated) */
+export const parse = (template, parent) => {
+  if (!parent) {
+    parent = new RootNode();
+  }
```

Comment:
제가 너무 기계적으로 리뷰를 하고 있다는 생각이 들지만... 그냥 가볍게 한번 봐주시면 좋을 거 같아요 :smiley: 


```suggestion
export const parse = (template, parent  = new RootNode()) => {
```

또는 함수 호출하는 쪽에서 명시적으로 RootNode를 인자로 넘기는건 어떨까요?

`!parent` 가 `true`가 되는 상황은 기계적으로 보면 parent가 `0`일 때, `""`일 때, `NaN`일 때도 가능하기 때문에,

명시적으로 `parent === undefined || parent === null` 라고 하는 건 어떨지 가볍게 말씀드려봅니다 :cat: 

참고로 저는 `parent == null` 이라고 적는걸 선호합니다 ㅎㅎ

Reply:
전혀 기계적이지 않고 제가 놓쳤던 부분 꼼꼼하게 봐주셔서 감사하게 보고 있습니다~!!

Reply:
`class.js`에 정의해둔 `RootNode`, `Node`, `TextNode` 등은 Virtual 노드로서, 메인 로직이 있는 `index.js`에서는 이 Virtual 노드 하나하나의 존재는 신경쓰거나 직접 생성하지 않기를 바랐어요. parser.js 모듈에서만 존재하면서 알아서(?) 처리되도록,,,, 그래서 호출하는 쪽에서 넘기는 대신 parse 함수 안에서 작성했었습니당. 이게 좋은 논리인지는 모르겠지만,,

이런 측면에서 제게는 동동이 제안해주신 방법 중 첫번째 방법이 좋아보여서 그대로 반영해보았습니다 :)
NaN이 falsy 값이라는 사실도 잊고 있었는데 덕분에 다시금 상기할 수 있었고, `parent === undefined || parent === null`를 합쳐서 `parent == null` 로 표현할 수 있다는 것도 새로 배웠습니다~!
좋은 피드백 감사합니다 😆

Reply:
📌 관련커밋: a79adba